Generic AAE Status

rzezeski opened the pull request:
This is replacing #596.  It's been rebased against develop.

This pull request makes AAE entropy info parameterized on type.  This allows both KV and Yokozuna to share almost all the same status code without copy pasta.
- Add a `Type` to entropy info.  It represents the type of AAE being performed.  Currently `riak_kv` or `yokozuna` but could be other things in the future.
- Export the `get_build_time` function as it can be reused by Yokozuna.
- Parameterize the module and function name used to compute all exchanges.  KV has inter-partition exchange whereas Yokozuna is purely intra.  I.e. Yokozuna exchange is always between the same partition, different services.  By parameterizing the "all exchanges" calculation can let different services have different exchange patterns.

rzezeski committed:
Add type to entropy info

Adding a type namespace to entropy info allows me to reuse it for
Yokozuna AAE status as well as future uses of AAE.

rzezeski committed:
Export get_build_time

This function can be reused by Yokozuna.

rzezeski committed:
Parametrize mod:fun to determine all exchanges

The exchanges that need to take place in Yokozuna are not the same as
those in KV.

rzezeski committed:
Fix AAE filtering

Fix the filtering logic which removes AAE info for indices no longer
owned by the local node.  The key is now `{Type, Index}`.

jrwest commented on src/riak_kv_entropy_info.erl:
```diff
@@ -132,10 +141,10 @@ all_index_info() ->
     Others = ordsets:subtract(ordsets:from_list(Indices),
                               ordsets:from_list(Primaries)),
-    [ets:delete(?ETS, {index, Idx}) || Idx <- Others],
+    [ets:match_delete(?ETS, {{index, {'_', Idx}}, '_'}) || Idx <- Others],
```
moving comments from https://github.com/basho/riak_kv/pull/596#discussion_r6211971 here:

I'm fine w/ the quick fix for now. I think we should open an issue as well since the table size does seem related to ring size (number of partitions owned by a single node). May not be that much of a hit vs. calling multiple deletes but since we did the large ring work we should be careful.

rzezeski replied:
https://github.com/basho/riak_kv/issues/657

rzezeski committed:
Make sure AAE status is backwards compatible

Add functions which provide a default `Type` so that the riak_kv calls
can stay as they currently are.

jrwest commented:
Changes look good. As mentioned in chat I'm +1 on this with the caveat that I think issues should be opened to export `tree_built/2` and to change to using `ets:delete` over `match_delete` (as pre2 issues). I didn't get a chance to run it with the changes but I can do that later today if we want to hold off a little longer on the merge.

jrwest commented on src/riak_kv_exchange_fsm.erl:
```diff
@@ -284,6 +284,7 @@ next_state_with_timeout(StateName, State, Timeout) ->
 
 exchange_complete({LocalIdx, _}, {RemoteIdx, RemoteNode}, IndexN, Repaired) ->
-    riak_kv_entropy_info:exchange_complete(LocalIdx, RemoteIdx, IndexN, Repaired),
+    riak_kv_entropy_info:exchange_complete(LocalIdx, RemoteIdx,
```
minor nitpick but do we even need this file in the patch now?

coderoshi replied:
:fish:

coderoshi committed:
Rollback unneeded file patch

rzezeski committed:
Add default type for tree_built

Use default type of `riak_kv` when calling `tree_built/2`.
